src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.proxy.listener;
+
+import io.r2dbc.proxy.core.MethodExecutionInfo;
+import io.r2dbc.proxy.core.ProxyEventType;
+import io.r2dbc.proxy.core.QueryExecutionInfo;
+import io.r2dbc.spi.Batch;
+import io.r2dbc.spi.Statement;
+
+import java.util.function.BiFunction;
+
+/**
+ * Listener interface that is called by proxy on {@link io.r2dbc.spi.Connection},
+ * {@link io.r2dbc.spi.Statement}, or {@link io.r2dbc.spi.Batch}.
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public interface ProxyExecutionListener {
+
+    /**
+     * Called before and after every invocation of methods.
+     *
+     * Default implementation of this method dispatches to {@link #beforeMethod(MethodExecutionInfo)} and
+     * {@link #afterMethod(MethodExecutionInfo)}. Thus overriding this method will stop calling them
+     * unless subclass calls {@code super.onMethodExecution(executionInfo);}.
+     *
+     * @param executionInfo query execution context
+     */
+    default void onMethodExecution(MethodExecutionInfo executionInfo) {
+        ProxyEventType eventType = executionInfo.getProxyEventType();
+        if (eventType == ProxyEventType.BEFORE_METHOD) {
+            beforeMethod(executionInfo);
+        } else if (eventType == ProxyEventType.AFTER_METHOD) {
+            afterMethod(executionInfo);
+        }
+    }
```

Comment:
And what if neither condition is met?

Reply:
Hm, the calling logic kinds of ensures event type is either `BEFORE_METHOD` or `AFTER_METHOD`.

Actually, this is one of the part I changed the way from datasource-proxy jdbc to r2dbc.

In dsp-jdbc, instead of having this dispatcher method, proxy logic directly calls
`beforeMethod()` or `afterMethod()`. So, probably that might be better style than dispatching in default method like here.

I'll remove this method and directly call before/after methods.